### PR TITLE
 API: Add DataFrame.droplevel

### LINKED ISSUE #20342: API: Add DataFrame.droplevel
As a way to make dropping an index level possible in an method chain.

```python
>>> df = pd.DataFrame(np.arange(12).reshape(4, 3), columns=['a', 'b', 'c']).set_index(['a', 'b'])
>>> df.droplevel(0)
```

roughly equivalent to 

```python
df2 = df.copy()
df2.index = df.index.droplevel(0)
df2
```

cc @twiecki

### PR DESCRIPTION
**Pull request by vbarda**
- [x] closes #20342
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by datapythonista** [comment]
Thanks for the contribution @vbarda.

From the issue, I'd say we want this method in index, not in DataFrame. @TomAugspurger can you please confirm?

Also, you're missing adding a comment to the whatsnew documentation (if you don't know how to do it, you can check closed PRs to see how it's done).

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good. can you add:

- whatsnew note in 0.24.0, new features, with a reference to the method
- add the method in api.rst 
- move to generic as indicated.

**Comment by vbarda** [comment]
Thank you for the comments -- addressing the changes now

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -74,6 +74,7 @@ Other Enhancements
 - :func:`Series.mode` and :func:`DataFrame.mode` now support the ``dropna`` parameter which can be used to specify whether NaN/NaT values should be considered (:issue:`17534`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
 - :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with :class:`MultiIndex` (:issue:`21115`)
+- :meth:`Series.droplevel` and `DataFrame.droplevel` are now implemented
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: add the issue number here

File: pandas/core/generic.py [hunk 2]
```diff
@@ -716,6 +716,62 @@ def swapaxes(self, axis1, axis2, copy=True):
 
         return self._constructor(new_values, *new_axes).__finalize__(self)
 
+    def droplevel(self, level, axis=0):
+        """Return DataFrame with requested index / column level(s) removed.
+        .. versionadded:: 0.24.0
```
Comment (jreback) [inline pandas/core/generic.py]: I think you need a blank line here? @jorisvandenbossche

File: pandas/core/generic.py [hunk 3]
```diff
@@ -716,6 +716,62 @@ def swapaxes(self, axis1, axis2, copy=True):
...
+
+        Parameters
+        ----------
+        level : int, str, or list-like
+            If a string is given, must be the name of a level
+            If list-like, elements must be names or indexes of levels.
```
Comment (jreback) [inline pandas/core/generic.py]: indexes -> positional indexes

File: pandas/tests/frame/test_alter_axes.py [hunk 4]
```diff
@@ -1056,6 +1056,38 @@ def test_reindex_signature(self):
                               "limit", "copy", "level", "method",
                               "fill_value", "tolerance"}
 
+    def test_droplevel(self):
+        # GH20342
+        df = pd.DataFrame([
+            [1, 2, 3, 4],
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: can you add tests for series as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_alter_axes.py (hunk 4 above)
Comment (vbarda) [reply pandas/tests/frame/test_alter_axes.py]: I did, on lines 1085-1089 :)

Comment (jreback) [reply pandas/tests/frame/test_alter_axes.py]: oh not in THIS test. pls separate out and put in tests/series/test_alter_axes

File: pandas/tests/frame/test_alter_axes.py [hunk 5]
```diff
@@ -1056,6 +1056,38 @@ def test_reindex_signature(self):
...
+            [5, 6, 7, 8],
+            [9, 10, 11, 12]
+        ])
+        df = df.set_index([0, 1]).rename_axis(['a', 'b'])
+        ser = df.iloc[:, 0]
+        df.columns = pd.MultiIndex.from_tuples([('c', 'e'), ('d', 'f')],
+                                               names=['level_1', 'level_2'])
+
+        # test that dropping of a level in DataFrame index works
+        expected_df_no_level_a_in_index = df.reset_index('a', drop=True)
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: dont' use these long names, expected and result are much easier to grok

File: pandas/tests/frame/test_alter_axes.py [hunk 6]
```diff
@@ -1056,6 +1056,38 @@ def test_reindex_signature(self):
...
+        expected_df_no_level_2_in_columns = df.copy()
+        expected_df_no_level_2_in_columns.columns = pd.Index(['c', 'd'],
+                                                             name='level_1')
+        actual_df_no_level_2_in_columns = df.droplevel('level_2', axis=1)
+        assert_frame_equal(expected_df_no_level_2_in_columns,
+                           actual_df_no_level_2_in_columns)
+
+        # test that dropping of a level in Series index works
+        expected_ser_no_level_b_in_index = ser.reset_index('b', drop=True)
+        actual_ser_no_level_b_in_index = ser.droplevel('b')
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: when you move add a test that raises when axis !=0 is specified

File: pandas/tests/frame/test_alter_axes.py [hunk 7]
```diff
@@ -1056,6 +1056,38 @@ def test_reindex_signature(self):
...
+        df.columns = pd.MultiIndex.from_tuples([('c', 'e'), ('d', 'f')],
+                                               names=['level_1', 'level_2'])
+
+        # test that dropping of a level in DataFrame index works
+        expected_df_no_level_a_in_index = df.reset_index('a', drop=True)
+        actual_df_no_level_a_in_index = df.droplevel('a')
+        assert_frame_equal(expected_df_no_level_a_in_index,
+                           actual_df_no_level_a_in_index)
+
+        # test that dropping of a level in DataFrame columns works
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: can you parameterize these on axis and also try with axis='columns',  and axis='index'

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt [hunk 8]
```diff
@@ -74,6 +74,7 @@ Other Enhancements
 - :func:`Series.mode` and :func:`DataFrame.mode` now support the ``dropna`` parameter which can be used to specify whether NaN/NaT values should be considered (:issue:`17534`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
 - :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with :class:`MultiIndex` (:issue:`21115`)
+- :meth:`Series.droplevel` and `DataFrame.droplevel` are now implemented (:issue:`20342`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: need ref for DataFrame.droplevel

File: pandas/core/generic.py [hunk 9]
```diff
@@ -716,6 +716,64 @@ def swapaxes(self, axis1, axis2, copy=True):
...
+        DataFrame.droplevel()
+
+        Examples
+        --------
+        >>> df = pd.DataFrame([
+        ...:     [1, 2, 3, 4],
+        ...:     [5, 6, 7, 8],
+        ...:     [9, 10, 11, 12]
+        ...: ]).set_index([0, 1]).rename_axis(['a', 'b'])
+        >>> df.columns = pd.MultiIndex.from_tuples([
```
Comment (jreback) [inline pandas/core/generic.py]: can you add blank lines in between here

File: pandas/core/generic.py [hunk 10]
```diff
@@ -716,6 +716,64 @@ def swapaxes(self, axis1, axis2, copy=True):
 
         return self._constructor(new_values, *new_axes).__finalize__(self)
 
+    def droplevel(self, level, axis=0):
```
Comment (jreback) [inline pandas/core/generic.py]: anyone think we should add ``inplace`` to make this consistent, with other axis ops (e.g. reset_index / set_index)?

cc @TomAugspurger

**Review by jreback — CHANGES_REQUESTED** [review]
@vbarda minor comments, but have a question about the API.

File: doc/source/whatsnew/v0.24.0.txt (hunk 8 above)
Comment (vbarda) [reply doc/source/whatsnew/v0.24.0.txt]: could you please explain more, not sure what you mean

File: pandas/core/generic.py (hunk 10 above)
Comment (vbarda) [reply pandas/core/generic.py]: makes sense to me to expose `inplace`!

File: doc/source/whatsnew/v0.24.0.txt (hunk 8 above)
Comment (TomAugspurger) [reply doc/source/whatsnew/v0.24.0.txt]: You have

    `DataFrame.droplevel`

Should be

    :meth:`DataFrame.droplevel`

File: pandas/core/generic.py (hunk 10 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: I'd say leave it out for now, unless someone asks for it.

File: doc/source/whatsnew/v0.24.0.txt (hunk 8 above)
Comment (vbarda) [reply doc/source/whatsnew/v0.24.0.txt]: thank you!

File: pandas/core/generic.py (hunk 10 above)
Comment (vbarda) [reply pandas/core/generic.py]: okay, will do!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @vbarda  nice patch!

can you create an issue about adding ``inplace`` to ``.droplevel``.

**Comment by vbarda** [comment]
Thank you -- will add the issue now!

**Comment by vbarda** [comment]
Added here: https://github.com/pandas-dev/pandas/issues/21990

**Pull request merged**